api/projects/permissions.py

```diff
@@ -155,10 +159,40 @@ def has_permission(self, request, view):
         return view.detail
 
     def has_object_permission(self, request, view, obj):
+        if request.user.is_anonymous:
+            return False
+
         if view.action in self.action_permission_map:
             return request.user.has_project_permission(
                 self.action_permission_map[view.action],
                 self.get_project_from_object_callable(obj),
             )
 
         return request.user.is_project_admin(self.get_project_from_object_callable(obj))
+
+
+class NestedProjectMasterAPIKeyPermissions(BasePermission):
+    # NOTE: In order to compose different permissions using Python bitwise operators
+    # all the permissions must have same __init__ signature. The __init__ method defined below
+    # provides the same signature as the one defined in the NestedProjectPermissions class
+    # to support composition using bitwise operators.
+    def __init__(self, *args, **kwargs):
+        super().__init__()
+
+    def has_permission(self, request: HttpRequest, view: str) -> bool:
+        master_api_key = getattr(request, "master_api_key", None)
+
+        if not master_api_key:
+            return False
+
+        with suppress(Project.DoesNotExist):
+            project_id = view.kwargs.get("project_pk")
+            project = Project.objects.get(id=project_id)
+
+            return project.organisation_id == master_api_key.organisation_id
+        return False
+
+    def has_object_permission(
+        self, request: HttpRequest, view: str, obj: Project
+    ) -> bool:
+        return self.has_permission(request, view)
```

Comment:
A slight optimisation here could be to create a new private method that takes the project as an optional parameter, right? Since we'd likely already have the project at `obj.project` so we could avoid the extra hit to the DB. Perhaps it's not worth it but if it's not too much extra code I think it makes sense.

Reply:
No, in this case the object will be a `MultivariateFeatureOption` which does not have a project. A possible solution (which I think we will have to implement in the near future) might involve accepting something like `get_project_from_object_callable` but that will still need another query in this case.

Reply:
Yep, ok, makes sense